README
                  
                      THC-IPV6-ATTACK-TOOLKIT
                (c) 2005-2011 vh@thc.org www.thc.org


		Licensed under GPLv3 (see LICENSE file)


INTRODUCTION
============
This code was inspired when I got into touch with IPv6, learned more and
more about it - and then found no tools to play (read: "hack") around with.
First I tried to implement things with libnet, but then found out that
the ipv6 implementation is only partial - and sucks. I tried to add the
missing code, but well, it was not so easy, hence I saved my time and
quickly wrote my own library.


LIMITATIONS
===========
This code currently only runs on:
 - Linux 2.6.x (because of /proc usage)
 - 32 Bit
 - Ethernet and Raw are supported (is there anything else necessary?)
But this means for all linux guys that it will work for 98% of your use cases.
Patches are welcome! (add "antispam" in the subject line to get through my
anti-spam protection, otherwise the email will bounce)


THE TOOLS
=========
The THC IPV6 ATTACK TOOLKIT comes already with lots of effective attacking
tools:
 - parasite6: icmp neighbor solitication/advertisement spoofer, puts you
   as man-in-the-middle, same as ARP mitm (and parasite)
 - alive6: an effective alive scanng, which will detect all systems
   listening to this address
 - dnsdict6: parallized dns ipv6 dictionary bruteforcer
 - fake_router6: announce yourself as a router on the network, with the
   highest priority
 - redir6: redirect traffic to you intelligently (man-in-the-middle) with
   a clever icmp6 redirect spoofer
 - toobig6: mtu decreaser with the same intelligence as redir6
 - detect-new-ip6: detect new ip6 devices which join the network, you can
   run a script to automatically scan these systems etc.
 - dos-new-ip6: detect new ip6 devices and tell them that their chosen IP
   collides on the network (DOS).
 - trace6: very fast traceroute6 with supports ICMP6 echo request and TCP-SYN
 - flood_router6: flood a target with random router advertisements
 - flood_advertise6: flood a target with random neighbor advertisements
 - fuzz_ip6: fuzzer for ipv6
 - implementation6: performs various implementation checks on ipv6
 - implementation6d: listen daemon for implementation6 to check behind a FW
 - fake_mld6: announce yourself in a multicast group of your choice on the net
 - fake_mld26: same but for MLDv2
 - fake_mldrouter6: fake MLD router messages
 - fake_mipv6: steal a mobile IP to yours if IPSEC is not needed for authentication
 - fake_advertiser6: announce yourself on the network
 - smurf6: local smurfer
 - rsmurf6: remote smurfer, known to work only against linux at the moment
 - exploit6: known ipv6 vulnerabilities to test against a target
 - denial6: a collection of denial-of-service tests againsts a target
 - thcping6: sends a hand crafted ping6 packet
 - sendpees6: a tool by willdamn@gmail.com, which generates a neighbor
   solicitation requests with a lot of CGAs (crypto stuff ;-) to keep the
   CPU busy. nice.
Just run the tools without options and they will give you help and show the
command line options.



THE LIBRARY
===========
The library thc-ipv6-lib.c is the heart and soul of all tools - and those
you may want to write.
Implementation is so simple, its usually just 2-4 lines to create a complete
ipv6/icmp6 packet with the content of your choice.

Your basic structure you use is
  (thc_ipv6_hdr *)
e.g. 
  thc_ipv6_hdr *my_ipv6_packet;
  int my_ipv6_packet_len;
and you will never have to play with its options/fields.

Whenever you want to build an ip6 packet, you just write:
  my_ipv6_packet = thc_create_ipv6(interface, prefer, &my_ipv6_packet_len,
                          src6, dst6, ttl, length, label, class, version);
if something fails, it returns NULL (only if my_ipv6_packet_len or dst6 do
not exist or malloc fails).
The options to thc_create_ipv6 are:
 (char*) interface - the interface on which you want to send out the packet
 (int) prefer - either PREFER_LINK (to use the link local address) or
                PREFER_HOST to use a host ip6 address, and 
                PREFER_GLOBAL to use a public (internet) IP6 address (default)
 (int *) &my_ipv6_packet_len - the size of the packet which will be created
 (unsigned char*) src6 - the source IP6 (OPTIONAL - will be selected if NULL)
 (unsigned char*) dst6 - the destination IP6 (in network format, 16 bytes long)
                         usually the result of thc_resolve6("ipv6.google.com");
 (int) ttl - the ttl of the packet (OPTIONAL - 0 will set this to 255)
 (int) length - the length which will be set in the header (OPTIONAL - 0 =
                real length)
 (int) label - the flow label (0 is fine)
 (int) class - the class of the packet (0 is fine)
 (int) version - the IP6 version (OPTIONAL - 0 will set this to version 6)
It returns NULL on errors or a malloc'ed structure on success.
free() it once you are done with it.

Now you can set extension headers on top of it:
  thc_add_hdr_route(my_ipv6_packet, &my_ipv6_packet_len, routers, routerptr);
  thc_add_hdr_fragment(my_ipv6_packet, &my_ipv6_packet_len, offset, more_frags,
                       id);
  thc_add_hdr_dst(my_ipv6_packet, &my_ipv6_packet_len, buf, buflen);
  thc_add_hdr_hopbyhop(my_ipv6_packet, &my_ipv6_packet_len, buf, buflen);
  thc_add_hdr_nonxt(my_ipv6_packet, &my_ipv6_packet_len, hdropt);
  thc_add_hdr_misc(my_ipv6_packet, &my_ipv6_packet_len, type, len, buf, buflen);
The functions explained:
 _route: Add a Routing Forwarding Header (like IP Source Routing)
  (int) routers - the number of routers in routerptr
  (char**) routerptr - a *char[routers + 1] struct with router destinations
                       in network format. See alive6.c for an example.
 _fragment: Add a Fragment Header
  (int) offset - the offset on which to the data should be written (note:
                 put the offset location in bytes here, not in byte octets)
  (int) more_frags - set to 0 if it is the fragement, 1 on all others
  (int) id - an ID for the packet (same for all fragments)
 _dst: Add a Destination Options Header
  (char*) buf - a char buffer. you have to control this buffer yourself with
                but you want to write into it.
  (int) buflen - the length of buf
 _hopbyhop: Add a Hop-By-Hop Header
  (char*) buf - a char buffer. you have to control this buffer yourself with
                but you want to write into it.
  (int) buflen - the length of buf
 _nonxt: Specify that there will be no following headers whatsoever
  (int) hdropt - this options is currently ignored
 _misc: Specify a miscelleanous header. Use this if you want to design an
        invalid or non-existing extension header.
  (int) type - The type ID to specify the header as
  (int) len - The length to advertise the header as (OPTIONAL - -1 sets this to
              the correct value)
  (char*) buf - a char buffer. you have to control this buffer yourself with
                but you want to write into it.
  (int) buflen - the length of buf
These functions return (int) 0 on success and -1 on error.

Finally you can add the stream or dgram headers.
  thc_add_icmp6(my_ipv6_packet, &my_ipv6_packet_len, type, code, flags, buf,
                buflen, checksum);
  thc_add_tcp(my_ipv6_packet, &my_ipv6_packet_len, source_port,
              destination_port, sequence_number, ack_number, flags, window_size
              urgent_pointer, options, optione_length, data, data_length);
  thc_add_udp(my_ipv6_packet, &my_ipv6_packet_len, source_port, 
              destination_port, checksum, data, data_length);
  thc_add_data6(my_ipv6_packet, &my_ipv6_packet_len, type, buf, buflen); 
 _icmp6: Add an ICMP6 packet header
  (int) type: the ICMP6 type
  (int) code: the ICMP6 code
  (int) flags: the ICMP6 flags
  (char*) buf - a char buffer. you have to control this buffer yourself with
                but you want to write into it.
  (int) buflen - the length of buf
 _tcp|_udp: Add an TCP or UDP header
  (ushort) source_port: source port
  (ushort) destination_port: destination port
  (uint) sequence_number: TCP sequence number
  (uint) ack_number: TCP acknowledgement number
  (ushort) checksum: UDP checksum, 0 = generate checksum (for TCP the checksum is always calculated)
  (uchar) flags: TCP flags: TCP_SYN, TCP_ACK, TCP_FIN, TCP_RST, TCP_PSH, ...
  (uint) window_size: TCP window size
  (uint) urgent_pointer: TCP urgent pointer (usually 0)
  (char*) options: TCP options buffer, can be NULL
  (uint) options_length: the length of the TCP options buffer
  (char*) data: the data the protocol carries
  (uint) data_length: the length of the data buffer
 _data6: Add a miscellaneous header
  (int) type: the protocol ID
  (char*) buf - a char buffer. you have to control this buffer yourself with
                but you want to write into it.
  (int) buflen - the length of buf
These functions return (int) 0 on success and -1 on error.

Once you are done, you create and send the packet.
  thc_generate_pkt(interface, srcmac, dstmac, my_ipv6_packet,
                   &my_ipv6_packet_len);
  thc_send_pkt(interface, my_ipv6_packet, &my_ipv6_packet_len);
or combined into one function:
 thc_generate_and_send_pkt(interface, srcmac, dstmac, my_ipv6_packet,
                            &my_ipv6_packet_len);

 thc_generate_and_send_pkt: This generates the real and final IPv6 packet and
                           then sends it.
  (char*) interface - the interface to send the packet on
  (unsigned char*) srcmac - the source mac to use (in network format)
                            (OPTIONAL, the real mac is used if NULL)
  (unsigned char*) dstmac - the destination mac to use (in network format)
                            (OPTIONAL, the real mac is looked up if NULL)
The thc_generate_pkt and  thc_send_pkt together provide the same functionality.
You usually use these only if you do something like
  thc_generate_pkt(...);
  while(1) thc_send_pkt(...);
These functions return (int) 0 on success and -1 on error.

When you are done, free the memory with:
  thc_destroy_packet(my_ipv6_packet);

There are some important helper functions you will need:
  thc_resolve6(destinationstring);
    This resolves the IPv6 address or DNS name to an IPv6 network address.
    Use this for dst6 in thc_create_ipv6(). The result has to be free'd when
    not needed anymore.
  thc_inverse_packet(my_ipv6_packet, &my_ipv6_packet_len);
    This clever functions switches source and destination address, exchanges
    the ICMP header type (ECHO REQUEST -> ECHO REPLY etc.) and recalculates
    the checksum. If you dont have an idea what this might be useful for,
    go and play with your xbox :-)
  thc_ipv6_rawmode(mode)
    the integer value 0 disables sending packets in raw mode, a value of 1
    enables it. Be warned that all MAC stuff will result in a dummy mac
    once it is in mode == 1.

If you just want to do it very fast, there are some predefined icmp6 creator
functions which sends impc6 packets in just one line of code:
  thc_ping6(interface, src, dst, size, count);
  thc_neighboradv6(interface, src, dst, srcmac, dstmac, flags, target);
  thc_neighborsol6(interface, src, dst, target, srcmac, dstmac);
  thc_routeradv6(interface, src, dst, srcmac, default_ttl, managed, prefix,
                 prefixlen, mtu, lifetime);
  thc_routersol6(interface, src, dst, srcmac, dstmac);
  thc_toobig6(interface, src, srcmac, dstmac, mtu, my_ipv6_packet,
              my_ipv6_packet_len);
  thc_paramprob6(interface, src, srcmac, dstmac, code, pointer,
                 my_ipv6_packet, my_ipv6_packet_len);
  thc_unreach6(interface, src, srcmac, dstmac, icmpcode, my_ipv6_packet,
               my_ipv6_packet_len);
  thc_redir6(interface, src, srcmac, dstmac, newrouter, newroutermac,
             my_ipv6_packet, my_ipv6_packet_len);
  thc_send_as_fragment6(interface, src, dst, type, buf, buflen, frag_len);
These do what you expect them to do, so I am too lazy^H^H^H^H^Hbusy to
describe it in more details.

The following functions allocate memory for the result pointer, so remember
to free the result pointers from these functions once you do not need them
anymore:
  thc_ipv6_dummymac()
  thc_ipv62notation()
  thc_ipv62string()
  thc_string2ipv6()
  thc_string2notation()
  thc_resolve6()
  thc_get_own_ipv6()
  thc_get_own_mac()
  thc_get_multicast_mac()
  thc_get_mac()
  thc_lookup_ipv6_mac()
  thc_look_neighborcache()
  thc_generate_key()
  thc_generate_cga()
  thc_generate_rsa()

It helps a lot if you take a look at example usages. The best ones are
the tools from the thc-ipv6 package, especially implementation6.c and
fake_*6.c - have fun, and send back code, so the community can further
build on it.


DETECTION
=========
Most tools can easily be detected by an IDS or specialized detection software.
This is done on purpose to make rogue usage detection easier.
The tools either specify a fixed packet signature, or generically sniff for
packets (e.g. therefore also answering to icmp6 neighbor solitications which
are sent to a non-existing mac, and are therefore very easy to detect).
If you dont want this, change the code.


PATCHES, BUGS, HINTS, etc.
==========================
Send them to vh (at) thc (dot) org (and add "antispam" to the subject line)

Have fun!


CHANGES
CHANGELOG
=========

NOTE: More tools exist, but are only handed out to specific people
who develop ipv6 security/pentest tools themselves, or support the
thc-ipv6 toolkit development. If this matches *you* send me an
email to vh (at) thc (dot) org , with "thc-ipv6 antispam" in the
subject line.


v1.8 - PUBLIC
 ! I am in a good mood, only the alive26, the advanced alive scanner is
 ! not in the package, so enoy lot of new powerful tools, plus most existing
 ! ones have been enhanced. Now there are 40 tools! Have fun on the CCC Camp!
 + Upcoming: even better fake_router6, and final death stroke to RA guard /
 +           NDP security
 
 * added new tool sendpeesmp6 by Marcin Pohl
 * added new tool randicmp6 by ecore
 * added new tool ndpexhaust6 by Mario Fleischmann
 * added two alternate alive6/parasite6 tools by Fabricio Nogueira Buzeto
   and Carlos Botelho De Paula Filho, it can be found in the contrib/ directory
 * added new tool flood_solicitate6
 * added new tool kill_router6
 * added new tool fake_dnsupdate6
 * added new tool node_query6
 * added new tool dump_router6
 * added helper scripts extract_{network,host}s.sh
 * new in public version: fake_dhcps6
 * new in public version: flood_dhcpc6
 * new in public version: fake_dns6d
 * new in public version: fake_dnsupdate6
 * new in public version: fragmentation6 (fragmentation test tool)
 * fake_advertise6: added one more ND Security bypass (-D)
 * fake_router6:
     - added unicast reply to router solicitation requests
     - added one more ND Security bypass (-D)
 * parasite6:
     - added -R option to also inject the reverse route
     - added one more ND Security bypass (-D)
 * flood_router6: one more RA guard bypass (-D)
 * dnsdict6:
     - expanded dictionary by results from the ipv6 world day scanning
     - added IPv4 support for selfish reasons. I'm sorry! ;-)
 * thcping6:
     - -D renamed to -F
     - new -D/-H option to specify options in hopbyhop and destination headers
     - fragment header moved before other headers (except hop-by-hop)
 * speed improvements for flood_* tools
 * thc-ipv6-lib:
     - fixed class assignment in ipv6 packet creation
     - forgot some fclose()es thanks to mario fleischmann for reporting
     - first OS/X porting diff sent in by oskar (at) acm (dot) org, thanks!


v1.7 - PRIVATE
 * ...


v1.6 - PUBLIC
 * removed various tools for public release


v1.5 - PRIVATE
 * redir6: 
     - TTL enhancement by frederik(at)kriewitz(dot)eu
     - timing enhancement by me
 * toobig6: added TTL, timing and packet size enhancement
 * parasite6:
     - added -l (loop) option
     - ND security evasion added :-)
 * fake_advertise6:
     - added src ip option
     - added ND security evasion options
 * trace6:
     - added tunnel detection and identification mode (-t)
     - only up to the 13th hop was reported, fixed
     - added patch by Phillipe Langlois for -s sourceipv6 option
 * alive26:
     - print original dst ports for packet replies
     - print original dst ipv6 for icmp errors
     - if -p was specified, sending dst opt error pkt was not disabled
 * thcping6: rewrote thcping6 for more options and packet timing
 * fake_router6:
     - the interface MTU is used as default now
     - added RA guard evasion options
 * flood_router6: RA guard evasion options added :-)
 * frag_id_attack: lots of more tests and cmdline options
 * implementation6: more test cases
 * dnsdict6:
     - implemended 4 different wordlists in dnsdict6 (-s, -m, -l, -x switches)
     - better wildcard detection
     - added -d switch to dump IPv6 NS and MX information
     - added check for -t max
 * added comfortable dnsdictalive.sh script to dnsbrute+alivescan a domain
 * thc-ipv6-lib:
     - added thc_ipv62notation function
     - added thc_add_hdr_oneshotfragment function
     - fixed neighbor mac solicitation function for FreeBSD targets
     - better own ipv6 address selection
 * added usage of thc_ipv62notation function to all tools

v1.4 - December 2010 - PUBLIC
 * removed various tools for public release

v1.3 - PRIVATE
 * added covert_send6 and covert_send6d
 * added fake_dhcps6 - fake dhcp6 server
 * added flood_dhcpc6 - dhcp6 flooder
 * added fake_dns6 - fake dns server, serving only one ipv6 address :-)
 * added fake_mld26 (same as fake_mld6 but for MLDv2)
 * added flood_mld6 - flood network with mld messages
 * added flood_mld26 - flood network with mldv2 messages
 * added fake_mldrouter6 - fake an mld router
 * added flood_mldrouter6 - flood network with mld router messages
 * added exploit6 and the first test cases
 * added denial6 and the first test cases
 * added dos_mld.sh which disables outside multicast traffic to the local LAN
 * alive6: 
    - beautified alive ipv6 address output
    - added -i inputfile and -o outputfile options
    - added -M for mac enumeration mode (autoconfiguration address space)
    - added -D for dhcp6 enumeration mode
    - added range possibility, e.g. "alive6 eth0 2002:0-2:0-10"
    - added -s/-a-/-u TCP SYN/ACK and UDP alive scan mode
    - added -F firewall quick port setup mode (tcp-syn to 22, 25, 80, 443;
       udp dns request; tcp-ack to highport, ping and destination error)
    - added -p and -e icmp and error alive check modes
    - changed hop-by-hop error check to destination error check
    - alive reply type is now printed
    - printing now warnings if icmp destination unreachables are received (-v)
    - added new -S slow, -W waittime,-d resolve, -v verbose switches
    - added new -Z dstmac option
    - removed hop check, changed to dst hdr, and made it the default
    - removed memory leaks
 * fake_mld6:
    - added query and done MLD types
    - new command line option -l = loop
    - command line format changed
    - added target mac option, needed for new vulnerablity found
 * dnsdict6:
    - added 87 more entries to the dictionary
    - now identified even multiple wildcard IPs and displays them accordingly

thc-ipv6.h

/*
 * (c) 2011 by van Hauser / THC
 *
 * THC IPv6 Attack Library Header Files
 *
 */

#ifndef _THC_IPV6_H

#define _THC_IPV6_H

#include <pcap.h>
#include <openssl/rsa.h>

#define VERSION 	"v1.8"
#define AUTHOR 		"van Hauser / THC <vh@thc.org>"
#define RESOURCE	"www.thc.org"

// you can change the following if you want to:
#define SHOW_LIBRARY_ERRORS 1

//#define MULTICAST_ALL_NODES

#define ICMP6_UNREACH     1
#define ICMP6_TOOBIG      2
#define ICMP6_TTLEXEED    3
#define ICMP6_PARAMPROB   4
#define ICMP6_PING        128
#define ICMP6_PONG        129
#define ICMP6_PINGREQUEST 128
#define ICMP6_PINGREPLY   129
#define ICMP6_ECHOREQUEST 128
#define ICMP6_ECHOREPLY   129
#define ICMP6_MLD_QUERY   130
#define ICMP6_MLD_REPORT  131
#define ICMP6_MLD_DONE    132
#define ICMP6_ROUTERSOL   133
#define ICMP6_ROUTERADV   134
#define ICMP6_NEIGHBORSOL 135
#define ICMP6_NEIGHBORADV 136
#define ICMP6_REDIR       137
#define ICMP6_INFOREQUEST 139
#define ICMP6_INFOREPLY   140
#define ICMP6_INVNEIGHBORSOL 141
#define ICMP6_INVNEIGHBORADV 142
#define ICMP6_MLD2_REPORT 143
#define ICMP6_MOBILE_PREFIXSOL 146
#define ICMP6_MOBILE_PREFIXADV 147
#define ICMP6_CERTPATHSOL 148
#define ICMP6_CERTPATHADV 149
#define ICMP6_MLD_ROUTERADV 151
#define ICMP6_MLD_ROUTERSOL 152
#define ICMP6_MLD_ROUTERTERMINATION 153
#define ICMP6_ROUTERPROXYSOL 154
#define ICMP6_ROUTERPROXYADV 155


#define ICMP6_NEIGHBORADV_ROUTER   0x80000000
#define ICMP6_NEIGHBORADV_SOLICIT  0x40000000
#define ICMP6_NEIGHBORADV_OVERRIDE 0x20000000

#define PREFER_HOST     16
#define PREFER_LINK	32
#define PREFER_GLOBAL	 0

extern void thc_ipv6_show_errors(int mode);
extern int thc_pcap_function(char *interface, unsigned char *capture, char *function, char *opt);
extern pcap_t *thc_pcap_init(char *interface, unsigned char *capture);
extern pcap_t *thc_pcap_init_promisc(char *interface, unsigned char *capture);
extern void thc_ipv6_rawmode(int mode);
extern int thc_pcap_check(pcap_t * p, char *function, char *opt);
extern char *thc_pcap_close(pcap_t * p);
extern char *thc_resolve6(unsigned char *target);
extern unsigned char *thc_lookup_ipv6_mac(char *interface, unsigned char *dst);
extern unsigned char *thc_get_own_mac(char *interface);
extern int thc_get_mtu(char *interface);
extern unsigned char *thc_get_own_ipv6(char *interface, unsigned char *dst, int prefer);
extern unsigned char *thc_get_multicast_mac(unsigned char *dst);
extern unsigned char *thc_get_mac(char *interface, unsigned char *src, unsigned char *dst);
extern unsigned char *thc_inverse_packet(unsigned char *pkt, int pkt_len);
extern int thc_ping6(char *interface, unsigned char *src, unsigned char *dst, int size, int count);
extern int thc_neighboradv6(char *interface, unsigned char *src, unsigned char *dst, unsigned char *srcmac, unsigned char *dstmac, unsigned int flags, unsigned char *target);
extern int thc_neighborsol6(char *interface, unsigned char *src, unsigned char *dst, unsigned char *target, unsigned char *srcmac, unsigned char *dstmac);
extern int thc_routeradv6(char *interface, unsigned char *src, unsigned char *dst, unsigned char *srcmac, unsigned char default_ttl, int managed, unsigned char *prefix,
                          int prefixlen, int mtu, unsigned int lifetime);
extern int thc_routersol6(char *interface, unsigned char *src, unsigned char *dst, unsigned char *srcmac, unsigned char *dstmac);
extern int thc_toobig6(char *interface, unsigned char *src, unsigned char *srcmac, unsigned char *dstmac, unsigned int mtu, unsigned char *pkt, int pkt_len);
extern int thc_paramprob6(char *interface, unsigned char *src, unsigned char *srcmac, unsigned char *dstmac, unsigned char code, unsigned int pointer, unsigned char *pkt,
                          int pkt_len);
extern int thc_unreach6(char *interface, unsigned char *src, unsigned char *srcmac, unsigned char *dstmac, unsigned char code, unsigned char *pkt, int pkt_len);
extern int thc_redir6(char *interface, unsigned char *src, unsigned char *srcmac, unsigned char *dstmac, unsigned char *newrouter, unsigned char *newroutermac,
                      unsigned char *pkt, int pkt_len);
extern int thc_send_as_fragment6(char *interface, unsigned char *src, unsigned char *dst, unsigned char type, unsigned char *data, int data_len, int frag_len);
extern char *thc_create_ipv6(char *interface, int prefer, int *pkt_len, unsigned char *src, unsigned char *dst, int ttl, int length, int label, int class, int version);
extern int thc_add_hdr_misc(unsigned char *pkt, int *pkt_len, unsigned char type, int len, unsigned char *buf, int buflen);
extern int thc_add_hdr_route(unsigned char *pkt, int *pkt_len, unsigned char **routers, unsigned char routerptr);
extern int thc_add_hdr_mobileroute(unsigned char *pkt, int *pkt_len, unsigned char *dst);
extern int thc_add_hdr_oneshotfragment(unsigned char *pkt, int *pkt_len, unsigned int id);
extern int thc_add_hdr_fragment(unsigned char *pkt, int *pkt_len, int offset, char more_frags, unsigned int id);
extern int thc_add_hdr_dst(unsigned char *pkt, int *pkt_len, unsigned char *buf, int buflen);
extern int thc_add_hdr_hopbyhop(unsigned char *pkt, int *pkt_len, unsigned char *buf, int buflen);
extern int thc_add_hdr_nonxt(unsigned char *pkt, int *pkt_len, int hdropt);
extern int thc_add_icmp6(unsigned char *pkt, int *pkt_len, int type, int code, unsigned int flags, unsigned char *data, int data_len, int checksum);
extern int thc_add_tcp(unsigned char *pkt, int *pkt_len, unsigned short int sport, unsigned short int dport, unsigned int sequence, unsigned int ack, unsigned char flags,
                       unsigned short int window, unsigned short int urgent, char *option, int option_len, char *data, int data_len);
extern int thc_add_udp(unsigned char *pkt, int *pkt_len, unsigned short int sport, unsigned short int dport, unsigned int checksum, char *data, int data_len);
extern int thc_add_data6(unsigned char *pkt, int *pkt_len, unsigned char type, unsigned char *data, int data_len);
extern int thc_generate_and_send_pkt(char *interface, unsigned char *srcmac, unsigned char *dstmac, unsigned char *pkt, int *pkt_len);
extern int thc_generate_pkt(char *interface, unsigned char *srcmac, unsigned char *dstmac, unsigned char *pkt, int *pkt_len);
extern int thc_send_pkt(char *interface, unsigned char *pkt, int *pkt_len);
extern unsigned char *thc_destroy_packet(unsigned char *pkt);
extern int thc_open_ipv6();
extern int thc_is_dst_local(char *interface, unsigned char *dst);
extern int checksum_pseudo_header(unsigned char *src, unsigned char *dst, unsigned char type, unsigned char *data, int length);
extern int calculate_checksum(unsigned char *data, int data_len);
extern void thc_dump_data(unsigned char *buf, int len, char *text);
extern unsigned char *thc_ipv62string(unsigned char *ipv6);
extern unsigned char *thc_string2ipv6(unsigned char *string);
extern unsigned char *thc_string2notation(unsigned char *string);
extern unsigned char *thc_ipv62notation(unsigned char *string);
extern unsigned char *thc_memstr(char *haystack, char *needle, int haystack_length, int needle_length);
extern void thc_notation2beauty(unsigned char *ipv6);
extern int thc_bind_udp_port(int port);
extern int thc_bind_multicast_to_socket(int s, char *interface, char *src);

#define DO_CHECKSUM 0xfaf4

#define NXT_IP6 41

#define NXT_INVALID 128
#define NXT_IGNORE 31

#define NXT_HDR 0
#define NXT_HOP 0
#define NXT_HBH 0
#define NXT_ROUTE 43
#define NXT_FRAG 44
#define NXT_NONXT 59
#define NXT_OPTS 60
#define NXT_DST 60

#define NXT_ESP 50
#define NXT_AH 51

#define NXT_MIPV6 135
#define NXT_ICMP6 58
#define NXT_TCP 6
#define NXT_UDP 17

#define NXT_DATA 255

#define IPV6_FRAME_TYPE 0x86dd

#define TCP_CWR 128
#define TCP_ECN 64
#define TCP_URG 32
#define TCP_ACK 16
#define TCP_PSH 8
#define TCP_RST 4
#define TCP_SYN 2
#define TCP_FIN 1

typedef struct {
  unsigned char dst[6];
  unsigned char src[6];
  unsigned int type:16;
} thc_ethernet;


typedef struct {
  unsigned char *pkt;
  int pkt_len;
  char *next_segment;
  char *final;
  int final_type;
  unsigned int version;         // :4;
  unsigned char class;
  unsigned int label;           // :20;
  unsigned int length;          // :16;
  unsigned char next;
  unsigned char ttl;
  unsigned char src[16];
  unsigned char dst[16];
  unsigned char *final_dst;
  unsigned char *original_src;
} thc_ipv6_hdr;

typedef struct {
  char *next_segment;
  unsigned char next;
  unsigned char length;
  unsigned char *data;
  int data_len;
} thc_ipv6_ext_hdr;

typedef struct {
  unsigned char type;
  unsigned char code;
  unsigned int checksum:16;
  unsigned int flags;
  unsigned char *data;
  int data_len;
} thc_icmp6_hdr;

typedef struct {
  unsigned short int sport;
  unsigned short int dport;
  unsigned int sequence;
  unsigned int ack;
  unsigned char length;
  unsigned char flags;
  unsigned short int window;
  unsigned short int checksum;
  unsigned short int urgent;
  unsigned char *option;
  int option_len;
  unsigned char *data;
  int data_len;
} thc_tcp_hdr;

typedef struct {
  unsigned short int sport;
  unsigned short int dport;
  unsigned short int length;
  unsigned short int checksum;
  unsigned char *data;
  int data_len;
} thc_udp_hdr;

typedef struct {
  char *next_segment;
  char dummy[8];
} thc_dummy_hdr;

/*
typedef struct {
        unsigned int           nlmsg_len;
        unsigned short         nlmsg_type;
        unsigned short         nlmsg_flags;
        unsigned int           nlmsg_seq;
        unsigned int           nlmsg_pid;
} nlmsghdr;

typedef struct
{
        unsigned char           rtgen_family;
} rtgenmsg;

typedef struct {
  nlmsghdr nlh;
  rtgenmsg g;
} neigh_req;

typedef struct {
        unsigned short         nl_family;
        unsigned short         nl_pad;
        unsigned int           nl_pid;
        unsigned int           nl_groups;
} sockaddr_nl;
*/

typedef struct {
  unsigned char type;
  unsigned char len;
  unsigned char pad_len;
  unsigned char resv;

/* cga params */
  unsigned char modifier[16];
  unsigned char prefix[8];
  unsigned char collision_cnt;
  unsigned char coll2;
  unsigned char *pub_key;
  unsigned char *exts;

/* end of cga params */
  unsigned char *pad;
} thc_cga_hdr;

typedef struct {
  unsigned char type;
  unsigned char len;
  unsigned char resv[6];
  unsigned long long timeval;
} thc_timestamp_hdr;

typedef struct {
  unsigned char type;
  unsigned char len;
  char nonce[6];
} thc_nonce_hdr;

typedef struct {
  unsigned char type;
  unsigned char len;
  short int resv;
  unsigned char key_hash[16];
  char *sign;
  char *pad;
} thc_rsa_hdr;

typedef struct {
  RSA *rsa;
  int len;
} thc_key_t;

typedef struct {
  unsigned char *data;
  /* DER-encoded key length */
  int len;
} opt_t;

extern thc_key_t *thc_generate_key(int key_len);
extern thc_cga_hdr *thc_generate_cga(unsigned char *prefix, thc_key_t * key, unsigned char **cga);
extern int thc_add_send(unsigned char *pkt, int *pkt_len, int type, int code, unsigned int flags, unsigned char *data, int data_len, thc_cga_hdr * cga_hdr, thc_key_t * key,
                        unsigned char *tag, int checksum);

#endif
